include/z64save.h

```diff
@@ -24,6 +24,36 @@ typedef enum RespawnMode {
 
 #define SAVE_BUFFER_SIZE 0x4000
 
+typedef enum {
+    /* 0  */ MAGIC_STATE_IDLE, // Regular gameplay
+    /* 1  */ MAGIC_STATE_CONSUME_SETUP, // Sets the speed at which the magic border flashes
+    /* 2  */ MAGIC_STATE_CONSUME, // Consume magic until target is reached or no more magic is available
+    /* 3  */ MAGIC_STATE_METER_FLASH_1, // Flashes border
+    /* 4  */ MAGIC_STATE_METER_FLASH_2, // Flashes border and draws yellow magic to preview target consumption
+    /* 5  */ MAGIC_STATE_RESET, // Reset colors and return to idle
+    /* 6  */ MAGIC_STATE_METER_FLASH_3, // Flashes border with no additional behaviour
+    /* 7  */ MAGIC_STATE_CONSUME_LENS, // Magic slowly consumed by lens
+    /* 8  */ MAGIC_STATE_STEP_CAPACITY, // Step `magicCapacity` to full capacity
+    /* 9  */ MAGIC_STATE_FILL, // Add magic until magicFillTarget is reached
+    /* 10 */ MAGIC_STATE_CONSUME_GORON_ZORA_SETUP,
+    /* 11 */ MAGIC_STATE_CONSUME_GORON_ZORA, // Magic slowly consumed by goron spikes or zora shocks
+    /* 12 */ MAGIC_STATE_CONSUME_GIANTS_MASK // Magic slowly consumed by giants mask
+} MagicState;
+
+typedef enum {
+    /* 0 */ MAGIC_CONSUME_NOW, // Consume Magic immediately without preview
+    /* 1 */ MAGIC_CONSUME_WAIT_NO_PREVIEW, // Sets consume target but waits to consume. No yellow magic preview to target consumption. Unused
+    /* 2 */ MAGIC_CONSUME_NOW_ALT, // Identical behaviour to MAGIC_CONSUME_NOW. Unused
+    /* 3 */ MAGIC_CONSUME_LENS, // Lens consumption
+    /* 4 */ MAGIC_CONSUME_WAIT_PREVIEW, // Sets consume target but waits to consume. Draws yellow magic to target consumption
+    /* 5 */ MAGIC_CONSUME_GORON_ZORA, // Zora shock and Goron Spike Roll slow consumption
+    /* 6 */ MAGIC_CONSUME_GIANTS_MASK, // Giants Mask Slow consumption
+    /* 7 */ MAGIC_CONSUME_DEITY_BEAM // Fierce Deity Beam consumption, consumed magic now and not via request
+} MagicChangeType;
+
+#define MAGIC_NORMAL_METER 0x30
+#define MAGIC_DOUBLE_METER (2 * MAGIC_NORMAL_METER)
```

Comment:
I hope I'm not stumbling into a conversation that was already settled a long time ago, but I prefer magic to be in decimal rather than hexadecimal. The amount of magic that a magic arrow consumes is directly presented to the player:
![NKZE01_2022-07-25_01-47-44](https://user-images.githubusercontent.com/12245827/180737005-0252c097-50d2-4e9d-b283-103d4617a270.png)

If a unit of magic was just some weird internal thing that players weren't really aware of, I'd be totally fine keeping it as hex. But since the game actually just straight up says "a Fire Arrow consumes two units of magic", the player *could* figure out how much magic they have at any given point just by counting how many Fire Arrows they can shoot off before they run out of magic. And given that it's countable, I'd prefer decimal in this case.

Reply:
There is a significant problem with that, namely that the internal magic number is *not* the MP amount the game tells the player, it's actually double. To use your example, OoT Player has
```c
static u8 sMagicArrowCosts[] = { 4, 4, 8 };
```
Lens is so far the only thing that works with single numbers in this system.

Reply:
I did a test in MM to confirm these numbers, and yes: the numbers are the same as OoT.

i.e.
(MP2) = 4 units of magic in-game
(MP4) = 8 units of magic in-game

I don't necessarily have a preference to hex vs dec for magic. I don't see this MP2 being a reason to push dec over hex as it's just a single digit value that's half the in-game value of magic

The only real hard-coded magic values to begin with is:
`#define MAGIC_NORMAL_METER 0x30` for half a magic bar 
`gSaveContext.save.playerData.magic += 0x10;` for filling and extending the magic bar
`gSaveContext.save.playerData.magic--;` for some types of magic consumption
`gSaveContext.save.playerData.magic -= 2;` for other types of magic consumption
`gSaveContext.save.playerData.magic += 4;` for adding magic

Note that I've defined both double magic and magic jar refills in terms of `MAGIC_NORMAL_METER`

So all this question really begs is should we use
`#define MAGIC_NORMAL_METER 48`
`gSaveContext.save.playerData.magic += 16;`
instead, which I don't have a preference. Hex is slightly more even, and aligns with health, but it would only change two lines.

Reply:
MM doesn't have any big magic-using items. OoT has Nayru's Love, which uses 24 ("MP12")
```c
static u8 sMagicSpellCosts[] = { 12, 24, 24, 12, 24, 12 };
```
I suspect that the actual numbers were added as an afterthought, I don't think any other Zelda game gives actual numbers for magic use. I suspect this is going to be another "make a macro for external use" thing, which I think I mentioned before for hearts.

Reply:
Yeah I didn't realize what was presented to the user was any different from the internal representation. I'm fine with using hex then.